crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -825,19 +827,64 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_for_statement(&mut self, for_statement: &ast::StmtFor) {
         let ast::StmtFor {
             range: _,
-            target,
+            target: _,
             iter,
             body,
             orelse,
             is_async: _,
         } = for_statement;
 
         self.infer_expression(iter);
-        self.infer_expression(target);
+        self.infer_definition(for_statement);
         self.infer_body(body);
         self.infer_body(orelse);
     }
 
+    fn add_maybe_multiple_for_definitions(
+        &mut self,
+        for_target: &ast::Expr,
+        loop_var_ty: Type<'db>,
+        definition: Definition<'db>,
+    ) {
+        if let ast::Expr::Tuple(ast::ExprTuple { elts, .. })
+        | ast::Expr::List(ast::ExprList { elts, .. }) = for_target
+        {
+            for elt in elts {
+                // TODO(Alex): unpack `loop_var_ty` as well, in tandem with unpacking `for_target`
+                self.add_maybe_multiple_for_definitions(elt, loop_var_ty, definition);
+            }
+            return;
+        }
+
+        self.types
+            .expressions
+            .insert(for_target.scoped_ast_id(self.db, self.scope), loop_var_ty);
+        self.types.definitions.insert(definition, loop_var_ty);
+    }
```

Comment:
Eh, I guess this is quite a naive approach to unpacked definitions, since things like this are also possible:

```pycon
>>> for x, *y in [(1, 2, 3), (4, 5, 6)]:
...     print(f"{x=}, {y=}")
... 
x=1, y=[2, 3]
x=4, y=[5, 6]
```

Happy to take this out for now if we don't want it, since I know that there are already some TODOs in this file about unpacked assignments (and https://github.com/astral-sh/ruff/issues/12698 is open)

Reply:
Yeah I think we should probably remove this and follow the model of `infer_assignment_statement`, leaving unpacking as a TODO for now; we can tackle unpacking as a single project across all the definition kinds where it can occur.